query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/QueryRequestToPinotSQLConverter.java

```diff
@@ -262,61 +266,155 @@ private String convertOperator2String(Operator operator) {
+        builder.append(convertExpressionToString(rhs, paramsBuilder, executionContext));
+        builder.append(")");
+        break;
+      case CONTAINS_KEY:
+        LiteralConstant[] kvp = convertExpressionToMapLiterals(filter.getRhs(), filter.getLhs());
+        builder.append(
+            convertExpressionToMapKeyColumn(filter.getLhs(), this::isColumnMapAttribute));
+        builder.append(" = ");
+        builder.append(convertLiteralToString(kvp[MAP_KEY_INDEX], paramsBuilder));
+        break;
+      case CONTAINS_KEYVALUE:
+        kvp = convertExpressionToMapLiterals(filter.getRhs(), filter.getLhs());
+        String keyCol =
+            convertExpressionToMapKeyColumn(filter.getLhs(), this::isColumnMapAttribute);
+        String valCol =
+            convertExpressionToMapValueColumn(filter.getLhs(), this::isColumnMapAttribute);
+        builder.append(keyCol);
+        builder.append(" = ");
+        builder.append(convertLiteralToString(kvp[MAP_KEY_INDEX], paramsBuilder));
+        builder.append(" AND ");
+        builder.append(valCol);
+        builder.append(" = ");
+        builder.append(convertLiteralToString(kvp[MAP_VALUE_INDEX], paramsBuilder));
+        builder.append(" AND ");
+        builder.append(MAP_VALUE);
+        builder.append("(");
+        builder.append(keyCol);
+        builder.append(",");
+        builder.append(convertLiteralToString(kvp[MAP_KEY_INDEX], paramsBuilder));
+        builder.append(",");
+        builder.append(valCol);
+        builder.append(") = ");
+        builder.append(convertLiteralToString(kvp[MAP_VALUE_INDEX], paramsBuilder));
         break;
+      default:
+        rhs = handleValueConversionForLiteralExpression(filter.getLhs(), filter.getRhs());
+        builder.append(convertExpressionToString(filter.getLhs(), paramsBuilder, executionContext));
+        builder.append(" ");
+        builder.append(operator);
+        builder.append(" ");
+        builder.append(convertExpressionToString(rhs, paramsBuilder, executionContext));
     }
-    return "";
+    return builder.toString();
   }
 
-  private String convertExpressionToMapKeyColumn(Expression expression) {
-    if ((expression.getValueCase() == COLUMNIDENTIFIER)
-        || (expression.getValueCase() == ATTRIBUTE_EXPRESSION && !isComplexAttribute(expression))) {
+  private String handleFilterForMapAttribute(Filter filter, Builder paramsBuilder) {
+
+    if (!SUPPORTED_OPERATORS_FOR_MAP_ATTRIBUTES_WITH_SUBPATH.contains(filter.getOperator())) {
+      throw new UnsupportedOperationException(
+          "Unknown operator for map attributes:" + filter.getOperator());
+    }
+
+    LiteralConstant[] kvp = convertExpressionToMapLiterals(filter.getRhs(), filter.getLhs());
+    String keyCol = convertExpressionToMapKeyColumn(filter.getLhs(), this::isAttributeMapAttribute);
+    String valCol =
+        convertExpressionToMapValueColumn(filter.getLhs(), this::isAttributeMapAttribute);
+
+    return keyCol
```

Comment:
Why are there 3 clauses here? Doesn't this translate to something like
```SQL
tags__keys = 'span.kind' 
  AND tags__value = 'client' 
  AND MAP_VALUE(tags__keys, 'span.kind', tags__values) = 'client'
```

The first two are redundant, the condition should translate simply as 
```SQL
MAP_VALUE(tags__keys, 'span.kind', tags__values) = 'client'
```
That's the purpose of the whole design it allows the _only_ change in this whole file to be in the translation of an Expression representing a selection. If our diff is more than 20 lines, something's off.

Reply:
Taking this offline.

Reply:
resolved